src/connector/C#/src/test/FunctionTest/SubscribeAsync.cs

```diff
@@ -0,0 +1,241 @@
+            IntPtr subscribe = TDengine.Subscribe(conn, true, tableName, querySql, subscribeCallback1, IntPtr.Zero, 200);
+            Thread.Sleep(1000);
+            TDengine.Unsubscribe(subscribe, true);
+            UtilsTools.ExecuteUpdate(conn, insertSql2);
+            subscribe = TDengine.Subscribe(conn, false, tableName, querySql, subscribeCallback2, IntPtr.Zero, 200);
+            Thread.Sleep(1000);
+            TDengine.Unsubscribe(subscribe, false);
+            void SubCallback1(IntPtr subscribe, IntPtr taosRes, IntPtr param, int code)
+            {
+                if (code == 0 && taosRes != IntPtr.Zero)
+                {
+                    // cannot free taosRes using taosRes, otherwise will cause crash.
+                    UtilsTools.GetResDataWithoutFree(taosRes);
+                }
+                else
+                {
+                    output.WriteLine($"async query data failed, failed code {code}");
+                }
+
+            }
+
+            void SubCallback2(IntPtr subscribe, IntPtr taosRes, IntPtr param, int code)
+            {
+                if (code == 0 && taosRes != IntPtr.Zero)
+                {
+                    List<TDengineMeta> actualMeta = UtilsTools.GetResField(taosRes);
+                    List<String> actualResData = UtilsTools.GetResDataWithoutFree(taosRes);
+                    UtilsTools.DisplayRes(taosRes);
+                    if (actualResData.Count == 0)
+                    {
+                        output.WriteLine($"consume in subscribe callback without data");
+                    }
+                    else
+                    {
+                        output.WriteLine($"consume in subscribe callback with data");
+
+                        Assert.Equal(expectResData2.Count, actualResData.Count);
+                        output.WriteLine("Assert Meta data");
+                        //Assert Meta data
+                        for (int i = 0; i < actualMeta.Count; i++)
+                        {
+                            Assert.Equal(expectResMeta[i].name, actualMeta[i].name);
+                            Assert.Equal(expectResMeta[i].type, actualMeta[i].type);
+                            Assert.Equal(expectResMeta[i].size, actualMeta[i].size);
+                        }
+                        output.WriteLine("Assert retrieve data");
+                        // Assert retrieve data
+                        for (int i = 0; i < actualResData.Count; i++)
+                        {
+                            // output.WriteLine("index:{0},expectResData:{1},actualResData:{2}", i, expectResData[i], actualResData[i]);
+                            Assert.Equal(expectResData2[i].ToString(), actualResData[i]);
+                        }
+                    }
+                }
+                else
+                {
+                    output.WriteLine($"async query data failed, failed code {code}");
```

Comment:
how about "error code" here? and can we provide more info provided by taos_errstr() interface?

Reply:
fixed